pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from sumpy.kernel import (AxisTargetDerivative, AxisSourceDerivative,
+        KernelWrapper)
+
+from pymbolic.interop.sympy import PymbolicToSympyMapper, SympyToPymbolicMapper
+from pymbolic.mapper import Mapper
+from pymbolic.primitives import Product
+import sympy
+import functools
+from collections import defaultdict
+
+
+__all__ = (
+    "reduce_number_of_fmms",
+    )
+
+
+def reduce_number_of_fmms(int_gs, source_dependent_variables):
+    """
+    Reduce the number of FMMs needed for a system of
+    :class:`~pytential.symbolic.primitives.IntG` objects.
+
+    This is done by converting the ``IntG`` object to a matrix of polynomials
+    with d variables corresponding to d dimensions, where each variable represents
+    a (target) derivative operator along one of the axes. All the properties of
+    derivative operator that we want are reflected in the properties of the
+    polynomial including addition, multiplication and exact polynomial division.
+
+    This matrix is factored into two matrices, where the left hand side matrix
+    represents a transformation at the target, and the right hand side matrix
+    represents a transformation at the source.
+
+    If the expressions given are not linear, then the input expressions are
+    returned as is.
+
+    :arg int_gs: list of ``IntG`` objects.
+
+    :arg source_dependent_variables: list of :class:`pymbolic.primtives.Expression`
+        objects. When reducing FMMs, consider only these variables as dependent
+        on source. For eg: densities, source derivative vectors.
+
+    Note: there is no argument for target-dependent variables as the algorithm
+    assumes that there are no target-dependent-variables passed to this function.
+    (where a "target-dependent variable" represents a function discretized on the
+    targets)
+    """
+
+    dim = int_gs[0].target_kernel.dim
+    axis_vars = sympy.symbols(f"_x0:{dim}")
+
+    # A high level driver for this function should send int_gs that are common.
```

Comment:
"are common" is odd phrasing... "share certain properties"?

Reply:
Fixed in 35d63f5a0fc8a0fd4af16b35dc4f08c4d06b8f57